metasrv/src/watcher/watcher_manager.rs

```diff
@@ -0,0 +1,242 @@
+    pub key: String,
+    pub prev: Option<SeqV>,
+    pub current: Option<SeqV>,
+}
+
+#[derive(Clone, Debug)]
+pub struct WatcherStateMachineSubscriber {
+    event_tx: mpsc::UnboundedSender<WatcherEvent>,
+}
+
+#[derive(Clone)]
+pub enum WatcherEvent {
+    CreateWatcherEvent(CreateWatcherEvent),
+    StateMachineKvDataEvent(StateMachineKvData),
+    ShutdownEvent(()),
+}
+
+#[derive(Debug)]
+pub struct WatcherManager {
+    event_tx: mpsc::UnboundedSender<WatcherEvent>,
+
+    pub subscriber: WatcherStateMachineSubscriber,
+}
+
+struct WatcherManagerCore {
+    event_rx: mpsc::UnboundedReceiver<WatcherEvent>,
+
+    watcher_streams: BTreeMap<WatcherId, WatcherStream>,
+
+    /// map range to WatcherId
+    watcher_range_set: RangeSet<String, WatcherKey>,
+
+    current_watcher_id: WatcherId,
+}
+
+impl WatcherManager {
+    pub fn create() -> Self {
+        let (event_tx, event_rx) = mpsc::unbounded_channel();
+
+        let core = WatcherManagerCore {
+            event_rx,
+            watcher_streams: BTreeMap::new(),
+            watcher_range_set: RangeSet::new(),
+            current_watcher_id: 1,
+        };
+
+        let _h = tokio::spawn(core.watcher_manager_main());
+
+        WatcherManager {
+            event_tx: event_tx.clone(),
+            subscriber: WatcherStateMachineSubscriber { event_tx },
+        }
+    }
+
+    pub fn stop(&self) {
+        let _ = self.event_tx.send(WatcherEvent::ShutdownEvent(()));
+    }
+
+    pub fn create_watcher_stream(&self, request: WatchRequest, tx: WatcherStreamSender) {
+        let create: CreateWatcherEvent = (request, tx);
+        let _ = self.event_tx.send(WatcherEvent::CreateWatcherEvent(create));
+    }
+}
+
+impl WatcherManagerCore {
+    #[tracing::instrument(level = "trace", skip(self))]
+    async fn watcher_manager_main(mut self) {
+        loop {
+            tokio::select! {
+                event = self.event_rx.recv() => {
+                    if let Some(event) = event {
+                        match event {
+                            WatcherEvent::CreateWatcherEvent((req, tx)) => {self.create_watcher_stream(req, tx).await;},
+                            WatcherEvent::StateMachineKvDataEvent(kv) => {self.notify_event(kv).await;},
+                            WatcherEvent::ShutdownEvent(_) => {
+                                tracing::info!("watcher manager has been shutdown");
+                                break;
+                            }
+                        }
+                    }
+                },
+            }
+        }
+    }
+
+    #[tracing::instrument(level = "debug", skip(self))]
+    fn close_stream(&mut self, id: WatcherId) {
+        let watcher = self.watcher_streams.get(&id);
+        if let Some(watcher) = watcher {
+            let range = watcher.key.clone()..watcher.key_end.clone();
+            self.watcher_range_set.remove(range, WatcherKey {
+                id,
+                filter: FilterType::All,
+            });
```

Comment:
It looks like the filter should be part of the `WatcherStream`, but not part of the `WatchKey`.